File: src/test/java/com/moomark/post/controller/PostControllerTest.java
package com.moomark.post.controller;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

import com.moomark.post.configuration.passport.PassportResponse;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.moomark.post.PassportTestRepository;
import com.moomark.post.model.entity.Post;

import net.minidev.json.JSONObject;

@TestMethodOrder(value = MethodOrderer.OrderAnnotation.class)
@SpringBootTest
@AutoConfigureMockMvc
public class PostControllerTest {
  @Autowired
  private MockMvc mvc;

  @Autowired
  private ObjectMapper mapper;

  @Autowired
  private PassportTestRepository passportTestRepository;

  @Test
  @Order(1)
  public void getPostsCountBefore() throws Exception {
    long posts = Long.parseLong(mvc.perform(get("/api/v1/posts/count"))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString());

    assertEquals(posts, 0);
  }

  @Test
  @Order(2)
  public void writePost() throws Exception {
    String testTitle = "testTitle";
    String testContent = "testContent";

    JSONObject requestParams = new JSONObject();
    requestParams.put("title", testTitle);
    requestParams.put("content", testContent);

    PassportResponse passport = passportTestRepository.generatePassport();

    Post post = mapper.readValue(mvc
        .perform(post("/api/v1/post")
            .header("Content-Type", "application/json")
            .header("x-moom-passport-user", passport.getPassport())
            .header("x-moom-passport-key", passport.getKey())
            .content(requestParams.toJSONString()))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), Post.class);

    assertEquals(post.getTitle(), testTitle);
    assertEquals(post.getContent(), testContent);
    assertEquals(post.getRecommendCount(), 0);
    assertEquals(post.getViewsCount(), 0);
  }

  @Test
  @Order(3)
  public void getPostsCountAfter() throws Exception {
    long posts = Long.parseLong(mvc.perform(get("/api/v1/posts/count"))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString());

    assertEquals(posts, 1);
  }

  @Test
  @Order(4)
  public void getPosts() throws Exception {
    Post[] post = mapper.readValue(mvc
        .perform(get("/api/v1/posts").queryParam("limit", "1"))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), Post[].class);

    assertEquals(post.length, 1);
  }

  @Test
  @Order(5)
  public void getPost() throws Exception {
    String testTitle = "testTitleInfo";
    String testContent = "testContentInfo";

    JSONObject requestParams = new JSONObject();
    requestParams.put("title", testTitle);
    requestParams.put("content", testContent);
    PassportResponse passport = passportTestRepository.generatePassport();

    Post post = mapper.readValue(mvc
        .perform(post("/api/v1/post")
            .header("Content-Type", "application/json")
            .header("x-moom-passport-user", passport.getPassport())
            .header("x-moom-passport-key", passport.getKey())
            .content(requestParams.toJSONString()))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), Post.class);

    assertEquals(post.getTitle(), testTitle);
    assertEquals(post.getContent(), testContent);

    long postId = post.getId();

    Post resultPost = mapper.readValue(mvc
        .perform(get("/api/v1/post/" + postId)
            .header("Content-Type", "application/json"))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), Post.class);

    assertEquals(resultPost.getId(), postId);
    assertEquals(resultPost.getTitle(), testTitle);
    assertEquals(resultPost.getContent(), testContent);
  }

  @Test
  @Order(6)
  public void writePostEmptyTitle() throws Exception {
    String testContent = "testContent";

    JSONObject requestParams = new JSONObject();
    requestParams.put("title", "");
    requestParams.put("content", testContent);
    PassportResponse passport = passportTestRepository.generatePassport();

    mvc.perform(post("/api/v1/post")
        .header("Content-Type", "application/json")
        .header("x-moom-passport-user", passport.getPassport())
        .header("x-moom-passport-key", passport.getKey())
        .content(requestParams.toJSONString()))
        .andExpect(status().isBadRequest()).andReturn().getResponse().getContentAsString();
  }

  @Test
  @Order(7)
  public void writePostEmptyContent() throws Exception {
    String testTitle = "testTitle";

    JSONObject requestParams = new JSONObject();
    requestParams.put("title", testTitle);
    requestParams.put("content", "");
    PassportResponse passport = passportTestRepository.generatePassport();

    mvc.perform(post("/api/v1/post")
        .header("Content-Type", "application/json")
        .header("x-moom-passport-user", passport.getPassport())
        .header("x-moom-passport-key", passport.getKey())
        .content(requestParams.toJSONString()))
        .andExpect(status().isBadRequest()).andReturn().getResponse().getContentAsString();
  }

}

File: src/test/kotlin/com/moomark/post/controller/PostControllerTest.kt
package com.moomark.post.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.moomark.post.model.entity.Post
import com.moomark.post.repository.PassportTestRepository
import net.minidev.json.JSONObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@TestMethodOrder(value = MethodOrderer.OrderAnnotation::class)
@SpringBootTest
@AutoConfigureMockMvc
class PostControllerTest {
    @Autowired
    private lateinit var mvc: MockMvc

    @Autowired
    private lateinit var mapper: ObjectMapper

    @Autowired
    private lateinit var passportTestRepository: PassportTestRepository

    @Test
    @Order(1)
    fun getPostsCountBefore() {
        val posts =
            mvc
                .perform(get("/api/v1/posts/count"))
                .andExpect(status().isOk)
                .andReturn()
                .response
                .contentAsString
                .toLong()

        assertEquals(posts, 0L)
    }

    @Test
    @Order(2)
    fun writePost() {
        val testTitle = "testTitle"
        val testContent = "testContent"

        val requestParams = JSONObject()
        requestParams["title"] = testTitle
        requestParams["content"] = testContent

        val passport = passportTestRepository.generatePassport()

        val post =
            mapper.readValue(
                mvc
                    .perform(
                        post("/api/v1/post")
                            .header("Content-Type", "application/json")
                            .header("x-moom-passport-user", passport!!.passport)
                            .header("x-moom-passport-key", passport!!.key)
                            .content(requestParams.toJSONString())
                    ).andExpect(status().isOk)
                    .andReturn()
                    .response
                    .contentAsString,
                Post::class.java
            )

        assertEquals(post.title, testTitle)
        assertEquals(post.content, testContent)
        assertEquals(post.recommendCount, 0L)
        assertEquals(post.viewsCount, 0L)
    }

    @Test
    @Order(3)
    fun getPostsCountAfter() {
        val posts =
            mvc
                .perform(get("/api/v1/posts/count"))
                .andExpect(status().isOk)
                .andReturn()
                .response
                .contentAsString
                .toLong()

        assertEquals(posts, 1L)
    }

    @Test
    @Order(4)
    fun getPosts() {
        val post =
            mapper.readValue(
                mvc
                    .perform(get("/api/v1/posts").queryParam("limit", "1"))
                    .andExpect(status().isOk)
                    .andReturn()
                    .response
                    .contentAsString,
                Array<Post>::class.java
            )

        assertEquals(post.size, 1)
    }

    @Test
    @Order(5)
    fun getPost() {
        val testTitle = "testTitleInfo"
        val testContent = "testContentInfo"

        val requestParams = JSONObject()
        requestParams["title"] = testTitle
        requestParams["content"] = testContent

        val passport = passportTestRepository.generatePassport()

        val post =
            mapper.readValue(
                mvc
                    .perform(
                        post("/api/v1/post")
                            .header("Content-Type", "application/json")
                            .header("x-moom-passport-user", passport!!.passport)
                            .header("x-moom-passport-key", passport!!.key)
                            .content(requestParams.toJSONString())
                    ).andExpect(status().isOk)
                    .andReturn()
                    .response
                    .contentAsString,
                Post::class.java
            )

        assertEquals(post.title, testTitle)
        assertEquals(post.content, testContent)

        val postId = post.id

        val resultPost =
            mapper.readValue(
                mvc
                    .perform(
                        get("/api/v1/post/$postId")
                            .header("Content-Type", "application/json")
                    ).andExpect(status().isOk)
                    .andReturn()
                    .response
                    .contentAsString,
                Post::class.java
            )

        assertEquals(resultPost.id, postId)
        assertEquals(resultPost.title, testTitle)
        assertEquals(resultPost.content, testContent)
    }

    @Test
    @Order(6)
    fun writePostEmptyTitle() {
        val testContent = "testContent"

        val requestParams = JSONObject()
        requestParams["title"] = ""
        requestParams["content"] = testContent

        val passport = passportTestRepository.generatePassport()

        mvc
            .perform(
                post("/api/v1/post")
                    .header("Content-Type", "application/json")
                    .header("x-moom-passport-user", passport!!.passport)
                    .header("x-moom-passport-key", passport!!.key)
                    .content(requestParams.toJSONString())
            ).andExpect(status().isBadRequest)
            .andReturn()
            .response
            .contentAsString
    }

    @Test
    @Order(7)
    fun writePostEmptyContent() {
        val testTitle = "testTitle"

        val requestParams = JSONObject()
        requestParams["title"] = testTitle
        requestParams["content"] = ""

        val passport = passportTestRepository.generatePassport()

        mvc
            .perform(
                post("/api/v1/post")
                    .header("Content-Type", "application/json")
                    .header("x-moom-passport-user", passport!!.passport)
                    .header("x-moom-passport-key", passport!!.key)
                    .content(requestParams.toJSONString())
            ).andExpect(status().isBadRequest)
            .andReturn()
            .response
            .contentAsString
    }
}
